ForumSessionProvider/ForumSessionProvider.php

```diff
@@ -18,284 +18,851 @@
+
+		// Only SMF does this.
+		if ($this->FSSettings['NameStyle'] == 'smf')
+			$user->setOption('forum_member_id', $this->FSuser['id_member']);
+
+		$this->userWiki->setOption('forum_last_update', time());
+		$this->userWiki->saveSettings();
+	}
+
+	// !!! Add checks here for banning
+	private function isBanned()
+	{
+		switch ($this->wgFSPSoftware)
+		{
+			case 'elk1.0':
+			case 'elk1.1':
+				return false;
+				break;
+			case 'smf2.0':
+			case 'smf2.1':
+				return $this->isBannedSMF();
+				break;
+			default: return false;
+		}
+	}
+
+	private function isBannedSMF()
+	{
+		// Check their ban once every 5 minutes.
+		if (!(time() > ((int) $this->userWiki->getOption('forum_last_update_ban', 0) + 300)))
+			return $this->userWiki->getOption('forum_is_banned', 0);
+
+		$request = $this->db_query('
+			SELECT id_ban
+			FROM ' . $this->FSSettings['db_prefix'] . 'ban_items AS i
+			LEFT JOIN ' . $this->FSSettings['db_prefix'] . 'ban_groups AS g
+				ON (i.id_ban_group = g.id_ban_group)
+			WHERE i.id_member = ' . ( (int) $this->FSuser['id_member']) . '
+				AND (g.cannot_post = 1 OR g.cannot_login = 1)');
+
+		$banned = $this->db_num_rows($request);
+		$this->db_free_result($request);
+
+		$this->userWiki->setOption('forum_last_update_ban', time());
+		$this->userWiki->setOption('forum_is_banned', $banned);
+		$this->userWiki->saveSettings();
+
+		return $banned;
+	}
+
+	/**
+	 * Connect to the forum database.
+	 *
+	 * {@source}
+	 */
+	private function FSDBConnect()
+	{
+		switch ($this->wgFSPSoftware)
+		{
+			case 'elk1.0':
+			case 'elk1.1':
+				$this->db = new mysqli(
+					$this->FSSettings['db_server'],
+					$this->FSSettings['db_user'],
+					$this->FSSettings['db_passwd'],
+					$this->FSSettings['db_name']
+				);
+				$this->FSDBType = 'mysqli';
+				if (!$this->db->connect_error) {
+					$this->loaded = true;
+				}
+				else {
+					$this->logger->warning('MySQL Error: ' . $this->db->connect_error);
+				}
+				break;
+			case 'smf2.0':
+			case 'smf2.1':
+				return $this->FSDBConnectSMF();
+				break;
+			default: return false;
+		}		
+	}
+
+	/**
+	 * Connect to the database. Use the settings from smf.
+	 *
+	 * {@source}
+	 */
+	private function FSDBConnectSMF()
+	{
+		global $smf_settings, $smcFunc;
+
+		$this->FSDBType = !empty($this->FSSettings['db_type']) && file_exists($this->FSSettings['sourcedir'] . '/Subs-Db-' . $this->FSSettings['db_type'] . '.php')  ? $this->FSSettings['db_type'] : 'mysql';
+
+		if (!defined('SMF'))
+			define('SMF', 'WIKI');
+		if (!isset($smcFunc))
+			$smcFunc = array();
+
+		require_once($this->FSSettings['sourcedir'] . '/Subs-Db-' . $this->FSDBType . '.php');
+
+		$this->conn = smf_db_initiate(
+			$this->FSSettings['db_server'],
+			$this->FSSettings['db_name'],
+			$this->FSSettings['db_user'],
+			$this->FSSettings['db_passwd'],
+			$this->FSSettings['db_prefix']
+		);
+
+		// As of now, we don't suport anything other than UTF8 with SMF.
+		$smcFunc['db_query']('set_character_set', '

```

Comment:
I hacked this for Elliquiy anyway. Please do not assume this.

Reply:
Suggestions?  SMF does have a character set we could check for.  Its largely ignored in 2.1 since we migrated to full UTF-8.

Reply:
I wrote this before I wrote the DB layer comment, sorry. Should have removed this.